docs/source/optimisation.rst

```diff
@@ -147,6 +166,93 @@ A list of :code:`Callback`s to be executed each iteration can be passed to :code
 
 .. autoclass:: cil.utilities.callbacks.LogfileCallback
    :members:
+   
+
+Approximate gradient sum function 
+----------------------------------
+
+Alternatively, consider optimisation problems of the form: 
+
+.. math:: \sum_{i=0}^{n-1} f_{i} = (f_{0} + f_{1} + ... + f_{n-1})
+
+where :math:`n` is the number of functions.  Where there is a large number of :math:`f_i` or their gradients are expensive to calculate, stochastic optimisation methods could prove more efficient.   CIL provides an abstract base class which defines the sum function and overwrites the usual (full) gradient calculation with an approximate gradient. 
+
+The idea for this class and its sum functions is to consider that some stochastic optimisation algorithms can be viewed as deterministic gradient descent algorithms replacing the gradient with an approximate gradient. For example Stochastic Gradient Descent replaces the gradient in Gradient Descent with the gradient of just one of the :math:`f_i`. 
+ 
+CIL provides an abstract base class which defines the sum function and overwrites the usual (full) gradient calculation  of a sum function with an approximate gradient. Child classes of this abstract base class can define different approximate gradients with different mathematical properties.
+
+For example in the following table, the left hand column has the approximate gradient function subclass, the header row has the optimisation algorithm and the body of the table has the resulting stochastic algorithm.
+
++----------------+-------+------------+----------------+
+|                | GD    | ISTA       | FISTA          |
++----------------+-------+------------+----------------+
+| SGFunction     | SGD   | Prox-SGD   | Acc-Prox-SGD   |
++----------------+-------+------------+----------------+
+| SAGFunction\*  | SAG   | Prox-SAG   | Acc-Prox-SAG   |
++----------------+-------+------------+----------------+
+| SAGAFunction\* | SAGA  | Prox-SAGA  | Acc-Prox-SAGA  |
++----------------+-------+------------+----------------+
+| SVRGFunction\* | SVRG  | Prox-SVRG  | Acc-Prox-SVRG  |
++----------------+-------+------------+----------------+
+| LSVRGFunction\*| LSVRG | Prox-LSVRG | Acc-Prox-LSVRG |
++----------------+-------+------------+----------------+
+
+\*In development 
+
+The below is an example of Stochastic Gradient Descent built of the SGFunction and Gradient Descent algorithm:
+
+.. code-block :: python
```

Comment:
Open an issue to move this code example to a seperate place and then embed (and test) using sphinx